### PR TITLE
flink: extract sources/sinks from transformation tree

### PR DESCRIPTION
**Pull request by mobuchowski**
Currently, Flink integration makes visitors look at transformations - the internal flink representation of stream graph - to extract proper sources and sinks. This PR extracts responsibility of doing this to separate class, which only gathers sinks and lists of sources that are connected to this sink. We can ignore transformations happening in the middle, since Flink contains method to gather all transient "parent" transformations, so we get true lineage for particular sink - we won't assign wrong source in case we had disjointed job tree - which is possible in Flink.

Closes: https://github.com/OpenLineage/OpenLineage/issues/561

### CODE REVIEW

File: integration/flink/src/main/java/io/openlineage/flink/visitor/KafkaSinkVisitor.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    Field recordSerializerField = FieldUtils.getField(KafkaSink.class, "recordSerializer", true);
+    try {
+      KafkaRecordSerializationSchema<?> serializationSchema =
+          (KafkaRecordSerializationSchema<?>) recordSerializerField.get(kafkaSink);
+      Field topicSelectorField =
+          FieldUtils.getField(
+              serializationSchema.getClass().asSubclass(serializationSchema.getClass()),
+              "topicSelector",
+              true);
+      Field topicSelectorFunctionField =
```
Comment (pawel-big-lebowski) [inline integration/flink/src/main/java/io/openlineage/flink/visitor/KafkaSinkVisitor.java:40]: Could we create an integration tests which verifies an event emitted by a dummy emitter. 
I am not sure if this is required for this PR or can be done in a next one, but we need test coverage from the early beginning.

File: integration/flink/build.gradle [hunk 2]
```diff
@@ -138,7 +143,7 @@ def commonTestConfiguration = {
     systemProperties = [
             'junit.platform.output.capture.stdout': 'true',
             'junit.platform.output.capture.stderr': 'true',
-            'openlineage.flink.jar': "${archivesBaseName}-${project.version}.jar",
+            ' .flink.jar': "${archivesBaseName}-${project.version}.jar",
```
Comment (mr-yusupov) [inline integration/flink/build.gradle:146]: space and dot is left intentionally?

File: integration/flink/src/main/java/io/openlineage/flink/OpenLineageFlinkJobListener.java [hunk 3]
```diff
@@ -1,116 +1,146 @@
...
+  private final StreamExecutionEnvironment executionEnvironment;
+  private final OpenLineageContext openLineageContext;
 
+  public OpenLineageFlinkJobListener(StreamExecutionEnvironment executionEnvironment) {
+    makeTransformationsUnclearableList(executionEnvironment);
+    this.executionEnvironment = executionEnvironment;
+    this.openLineageContext =
+        OpenLineageContext.builder()
+            .streamExecutionEnvironment(Optional.of(executionEnvironment))
+            .openLineage(new OpenLineage(URI.create("")))
```
Comment (mr-yusupov) [inline integration/flink/src/main/java/io/openlineage/flink/OpenLineageFlinkJobListener.java:36]: I'm working on this URI part, trying to move code from spark side, I will think about making it common for both Flink/Spark modules

File: integration/flink/src/main/java/io/openlineage/flink/OpenLineageFlinkJobListener.java [hunk 4]
```diff
@@ -1,116 +1,146 @@
...
+      List<OpenLineage.OutputDataset> outputDatasets) {
+    OpenLineage.RunEvent runEvent =
+        openLineageContext
+            .getOpenLineage()
+            .newRunEventBuilder()
+            .inputs(inputDatasets)
+            .outputs(outputDatasets)
+            .build();
 
+    log.info("RunEvent: {}", runEvent);
```
Comment (mr-yusupov) [inline integration/flink/src/main/java/io/openlineage/flink/OpenLineageFlinkJobListener.java:79]: [Minor] maybe we should convert these logs to debug level?

File: integration/flink/src/main/java/io/openlineage/flink/TransformationConverter.java [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+ */
+@Slf4j
+public class TransformationConverter {
+
+  private final List<Object> sources = new ArrayList<>();
+  private final List<Object> sinks = new ArrayList<>();
+
+  private final Set<String> processedTransformations = new HashSet<>();
+
+  public List<SinkLineage> convertToVisitable(List<Transformation<?>> transformations) {
```
[the remark this answers is not in the document]
Comment (mobuchowski) [reply integration/flink/src/main/java/io/openlineage/flink/TransformationConverter.java:27]: Addressed this point offline with @pawel-big-lebowski .

To sum up, current structure is more decoupled - despite looking more ugly for experienced Java eyes, it allows us easier to change structure of how visitors or this transformation looks without influencing each other.

That said, I'm absolutely for changing it once the integration matures.

File: integration/flink/src/main/java/io/openlineage/flink/visitor/KafkaSinkVisitor.java (hunk 1 above)
Comment (mr-yusupov) [reply integration/flink/src/main/java/io/openlineage/flink/visitor/KafkaSinkVisitor.java:40]: I'm preparing emiitter (based on Spark emitter), and will create integration test for this case

File: integration/flink/src/main/java/io/openlineage/flink/visitor/KafkaSinkVisitor.java [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+  }
+
+  @Override
+  public boolean isDefinedAt(Object sink) {
+    return sink instanceof KafkaSink;
+  }
+
+  @Override
+  public List<OpenLineage.OutputDataset> apply(Object kafkaSink) {
+    var sink = (KafkaSink<?>) kafkaSink;
```
Comment (mr-yusupov) [inline integration/flink/src/main/java/io/openlineage/flink/visitor/KafkaSinkVisitor.java:30]: seems like we don't use this variable, can be removed

File: integration/flink/src/main/java/io/openlineage/flink/visitor/KafkaSinkVisitor.java (hunk 1 above)
Comment (mobuchowski) [reply integration/flink/src/main/java/io/openlineage/flink/visitor/KafkaSinkVisitor.java:40]: So, will be done in other PR.

File: integration/flink/src/main/java/io/openlineage/flink/visitor/KafkaSinkVisitor.java (hunk 6 above)
Comment (mobuchowski) [reply integration/flink/src/main/java/io/openlineage/flink/visitor/KafkaSinkVisitor.java:30]: Done.

File: integration/flink/build.gradle (hunk 2 above)
Comment (mobuchowski) [reply integration/flink/build.gradle:146]: Fixed.

File: integration/flink/src/main/java/io/openlineage/flink/OpenLineageFlinkJobListener.java [hunk 7]
```diff
@@ -1,116 +1,146 @@
...
+              FieldUtils.readField(transformations, executionEnvironment, true);
+      List<Transformation<?>> transformationList =
+          new UnclearableList<>(previousTransformationList);
+      FieldUtils.writeField(transformations, executionEnvironment, transformationList, true);
+    } catch (IllegalAccessException e) {
+      log.error("Failed to rewrite transformations");
+    }
+  }
+
+  static class UnclearableList<T> extends ArrayList<T> {
```
[the remark this answers is not in the document]
Comment (mobuchowski) [reply integration/flink/src/main/java/io/openlineage/flink/OpenLineageFlinkJobListener.java:135]: Done.

**Review by mr-yusupov — APPROVED** [review]

**Pull request merged**